tests/test_qpx.py

```
@@ -0,0 +1,320 @@
+            }
+        )
+
+    def test_counts_distinct_run_scan_pairs(self):
+        psm_df = self._psm_df()
+        # run1 has scans {1, 2} (2 rows share scan 2), run2 has scan {1} -> 3 pairs.
+        assert psm_df.groupby(["run", psm_df["scan"].str[0]]).ngroups == 3
+
+    def test_scan_column_is_not_directly_groupable(self):
+        psm_df = self._psm_df()
+        with pytest.raises(TypeError):
+            psm_df.groupby(["run", "scan"]).ngroups
```

Ruff (0.16.1)

[warning] Found useless attribute access. Either assign it to a variable or remove it.

(B018)